lesson-05-timer-management/timer_irq/xxx_irq.c

```diff
@@ -0,0 +1,96 @@
+#include <linux/module.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <asm/uaccess.h>
+#include <linux/pci.h>
+#include <linux/version.h>
+#include <linux/init.h>
+#include <linux/time.h>
+#include <linux/timer.h>
+
+
+#define LEN_MSG 160
+#define DELAY 1000
+static char* buf_msg;
+static struct timer_list irq_timer;
+static unsigned long time_interval;
+static ssize_t xxx_show( struct class *class, struct class_attribute *attr, char *buf ) {
+   
+   strcpy( buf, buf_msg );  
+   printk("timeout: %ld", time_interval);
+   printk( "read %ld\n", (long)strlen( buf ) );   
+   return strlen( buf );
+}
+
+static ssize_t xxx_store( struct class *class, struct class_attribute *attr, const char *buf, size_t count ) {
+   //printk( "write %ld\n", (long)count );
+   strncpy( buf_msg, buf, count );
+   buf_msg[ count ] = '\0';
+   return count;
+}
+
+static unsigned long get_fibonachi(unsigned long place)
+{
+       unsigned long old_value = 0;
+	unsigned long value = 1;
+	unsigned long hold;
+	unsigned long n;
+	if(value < 1) return (0);
+	for (n = 1; n < place; n++)
+	{
+		hold = value;
+		value += old_value;
+		old_value = hold;
+	}
+	return (value);
+}
+
+
+CLASS_ATTR_RW( xxx );
+
+static struct class *x_class;
+
+void IRQ_TimerHandler(unsigned long data)
+{
+	char buf[20];
+	time_interval ++;
+	sprintf(buf,"%u %llu\n", time_interval, get_fibonachi(time_interval));
+	xxx_store(x_class, &class_attr_xxx, buf, strlen(buf)+1);
```

Comment:
Interesting idea of reusing buffer functionality.
But when somebody writes to `x-class/xxx` something, following reading will return this string instead of expected fibonacci value until next second elapsed.